game/src/sandbox/mod.rs

```diff
@@ -662,3 +680,29 @@ impl ContextualActions for Actions {
         self.gameplay.clone()
     }
 }
+
+fn mouseover_unzoomed_agent_circle(ctx: &mut EventCtx, app: &mut App) {
+    // TODO How do we make this efficient?
+    // - We could build and cache a quadtree if we're paused
+    // - We could always build the quadtree as we loop over unzoomed agents
+    // - At least do this while we draw? Don't call twice.
```

Comment:
Very interesting! I've never worked with quadtrees, but they seem conceptually simple enough. 

My reading is that we're already using aabb_quadtree to index the static map elements: roads, buildings, intersections, etc, but not for anything dynamic.

And apparently we don't need this fancy data structure in the zoomed in version - is that because we can cheaply crop out agents not on the screen?

So, as you highlight, the big question is how expensive will it be to build/rebuild/update these for agents of interest. Good questions! I have no idea!

Reply:
One terrible hack idea: 

A quadtree is an index to quickly and repeatedly query into spatial data near arbitrary points. But for the specific problem of what's under the cursor, maybe we can simplify. We're only concerned with one point at any moment - the position of the map cursor. Would it be possible to approached the problem from the other angle, and instead of causing "hover" and "time passed" events to enumerate and compare against all agents, what if instead when we draw the agents, we checked the global "app.cursor_position" or something, and propagated some "is_hovered" property? 

Egregious, I know, but I had to mention it!

Reply:
> My reading is that we're already using aabb_quadtree to index the static map elements: roads, buildings, intersections, etc, but not for anything dynamic.

Correct

> And apparently we don't need this fancy data structure in the zoomed in version - is that because we can cheaply crop out agents not on the screen?

Check out `get_renderables_back_to_front`, by the comment `Expand all of the Traversables into agents, populating the cache if needed`. We use the static quadtree to figure out what lanes/turns are onscreen, then fetch the agents just on those, caching the `DrawCar`, `DrawPed`, etc.



I'm holding off figuring out how to wire in a quadtree, or plumb through the `is_hovered` check, for now and flag-guarding instead. There are a few other callers to `sim.get_unzoomed_agents` that also slow things down, but just when certain layers are open. I'm still puzzling through how to minimize these expensive calls and have different things "subscribe" to the call maybe.